File: leetcode/cpp/p67-add-binary.cpp
// LeetCode 67. Add Binary.
#include <iostream>

class Solution {
public:
    std::string addBinary(std::string a, std::string b)
    {
        std::string ans(std::max(a.size(), b.size()), '0');
        auto ans_it = ans.rbegin();
        auto ait = a.rbegin();
        auto bit = b.rbegin();
        bool carry = false;
        while (ait != a.rend() || bit != b.rend()) {
            const auto a_val = ait == a.rend() ? '0' : *ait++;
            const auto b_val = bit == b.rend() ? '0' : *bit++;
            const int total = static_cast<int>(a_val == '1') + static_cast<int>(b_val == '1') + static_cast<int>(carry);
            if (total == 1 || total == 3) {
                *ans_it = '1';
            }
            carry = total >= 2;
            ++ans_it;
        }
        if (carry) {
            return "1" + ans;
        }
        return ans;
    }

    // Nope, this didn't work.
    std::string addBinary2(std::string a, std::string b)
    {
        return std::to_string(std::stoi(a, nullptr, 2) + std::stoi(b, nullptr, 2));
    }
};

int main()
{
    struct TestCase {
        std::string a;
        std::string b;
        std::string exp;
    };

    TestCase test_cases[] = {
        { "0", "0", "0" },
        { "0", "1", "1" },
        { "1", "0", "1" },
        { "1", "1", "10" },
        { "10", "1", "11" },
        { "1", "10", "11" },
        { "11", "1", "100" },
        { "1", "11", "100" },
        { "0", "110", "110" },
        { "1010", "1011", "10101" },
        { "1011", "1010", "10101" },
    };
    Solution solution;
    for (auto& tc : test_cases) {
        const auto act = solution.addBinary(tc.a, tc.b);
        if (tc.exp != act) {
            std::cerr << "fail, a: " << tc.a
                      << ", b: " << tc.b
                      << ", exp: " << tc.exp
                      << ", act: " << act
                      << "\n";
        }
    }
}

File: leetcode/rust/p67-add-binary.rs
use std::cmp;

struct Solution {}

impl Solution {
    pub fn add_binary(a: String, b: String) -> String {
        let mut ans = String::with_capacity(cmp::max(a.len(), b.len()));
        let mut ait = a.chars().rev();
        let mut bit = b.chars().rev();
        let mut carry = 0;

        loop {
            let mut do_break = false;
            let mut total = carry;
            match ait.next() {
                Some(c) => {
                    if c == '1' {
                        total += 1;
                    }
                }
                None => do_break = true,
            }
            match bit.next() {
                Some(c) => {
                    if c == '1' {
                        total += 1;
                    }
                }
                None => {
                    if do_break {
                        break;
                    }
                }
            }
            if total == 1 || total == 3 {
                ans.push('1');
            } else {
                ans.push('0');
            }
            if total >= 2 {
                carry = 1;
            } else {
                carry = 0;
            }
        }

        if carry > 0 {
            ans.push('1');
        }
        return ans.chars().rev().collect::<String>();
    }
}

fn main() {
    struct TestCase {
        a: String,
        b: String,
        exp: String,
    }
    let test_cases = vec![
        TestCase {
            a: "0".to_string(),
            b: "0".to_string(),
            exp: "0".to_string(),
        },
        TestCase {
            a: "1".to_string(),
            b: "0".to_string(),
            exp: "1".to_string(),
        },
        TestCase {
            a: "1".to_string(),
            b: "1".to_string(),
            exp: "10".to_string(),
        },
        TestCase {
            a: "10".to_string(),
            b: "1".to_string(),
            exp: "11".to_string(),
        },
        TestCase {
            a: "1".to_string(),
            b: "10".to_string(),
            exp: "11".to_string(),
        },
        TestCase {
            a: "11".to_string(),
            b: "1".to_string(),
            exp: "100".to_string(),
        },
        TestCase {
            a: "1".to_string(),
            b: "11".to_string(),
            exp: "100".to_string(),
        },
        TestCase {
            a: "0".to_string(),
            b: "110".to_string(),
            exp: "110".to_string(),
        },
        TestCase {
            a: "1010".to_string(),
            b: "1011".to_string(),
            exp: "10101".to_string(),
        },
    ];

    for tc in test_cases.iter() {
        let act = Solution::add_binary(tc.a.clone(), tc.b.clone());
        if tc.exp != act {
            println!(
                "Fail, a: {:?}, b: {:?}, exp: {:?}, act: {:?}",
                tc.a, tc.b, tc.exp, act
            );
        }
    }
}
